File: statuserver/statuserver/server/storage.ts
import { 
  type Service, 
  type InsertService, 
  type Incident, 
  type InsertIncident,
  type StatusHistory,
  type InsertStatusHistory,
  type ServerMetrics,
  type InsertServerMetrics,
  type ServiceStatus
} from "@shared/schema";
import { randomUUID } from "crypto";
import { importServicesFromData } from "./importData";
import importedData from "./imported_services.json";

export interface IStorage {
  // Services
  getServices(): Promise<Service[]>;
  getService(id: string): Promise<Service | undefined>;
  createService(service: InsertService): Promise<Service>;
  updateServiceStatus(id: string, status: ServiceStatus): Promise<Service | undefined>;
  
  // Incidents
  getIncidents(): Promise<Incident[]>;
  getIncident(id: string): Promise<Incident | undefined>;
  createIncident(incident: InsertIncident): Promise<Incident>;
  
  // Status History
  getStatusHistory(serviceId: string): Promise<StatusHistory[]>;
  createStatusHistory(history: InsertStatusHistory): Promise<StatusHistory>;
  
  // Server Metrics
  getServerMetrics(serviceId?: string): Promise<ServerMetrics[]>;
  createServerMetrics(metrics: InsertServerMetrics): Promise<ServerMetrics>;
}

export class MemStorage implements IStorage {
  private services: Map<string, Service>;
  private incidents: Map<string, Incident>;
  private statusHistory: Map<string, StatusHistory>;
  private serverMetrics: Map<string, ServerMetrics>;

  constructor() {
    this.services = new Map();
    this.incidents = new Map();
    this.statusHistory = new Map();
    this.serverMetrics = new Map();
    
    // Seed with example data
    this.seedData();
  }

  private async seedData() {
    await importServicesFromData(this, importedData as any);
    
    return;
    
    const exampleServices: InsertService[] = [
      {
        name: "API Gateway",
        description: "Central API routing and management service",
        category: "App",
        region: "ru-central1-a",
        status: "operational",
        icon: "server",
      },
      {
        name: "Web Application",
        description: "Main web application frontend",
        category: "App",
        region: "ru-central1-a",
        status: "operational",
        icon: "globe",
      },
      {
        name: "Mobile API",
        description: "REST API for mobile applications",
        category: "App",
        region: "ru-central1-b",
        status: "degraded",
        icon: "server",
      },
      {
        name: "S3 Storage Bucket",
        description: "Object storage for static assets and user uploads",
        category: "services",
        region: "ru-central1-a",
        status: "operational",
        icon: "database",
      },
      {
        name: "GitLab Repository",
        description: "Source code management and CI/CD pipeline",
        category: "services",
        region: "ru-central1-c",
        status: "operational",
        icon: "server",
      },
      {
        name: "Redis Cache",
        description: "In-memory data structure store for caching",
        category: "services",
        region: "ru-central1-a",
        status: "operational",
        icon: "database",
      },
      {
        name: "Message Queue",
        description: "RabbitMQ message broker for async processing",
        category: "services",
        region: "ru-central1-b",
        status: "operational",
        icon: "server",
      },
      {
        name: "Elasticsearch",
        description: "Search and analytics engine",
        category: "services",
        region: "ru-central1-c",
        status: "maintenance",
        icon: "database",
      },
      {
        name: "Application Server 1",
        description: "Primary application server instance",
        category: "server",
        region: "ru-central1-a",
        status: "operational",
        icon: "server",
      },
      {
        name: "Application Server 2",
        description: "Secondary application server instance",
        category: "server",
        region: "ru-central1-a",
        status: "operational",
        icon: "server",
      },
      {
        name: "Database Server",
        description: "PostgreSQL database server",
        category: "server",
        region: "ru-central1-b",
        status: "operational",
        icon: "database",
      },
      {
        name: "Backup Server",
        description: "Automated backup and disaster recovery",
        category: "server",
        region: "ru-central1-c",
        status: "operational",
        icon: "shield",
      },
      {
        name: "Load Balancer",
        description: "Application load balancing across multiple instances",
        category: "server",
        region: "ru-central1-a",
        status: "operational",
        icon: "globe",
      },
      {
        name: "Monitoring Server",
        description: "System monitoring and alerting service",
        category: "server",
        region: "ru-central1-b",
        status: "down",
        icon: "server",
      },
    ];

    for (const service of exampleServices) {
      await this.createService(service);
    }

    // Create some example incidents
    const services = Array.from(this.services.values());
    if (services.length > 0) {
      const twoWeeksAgo = new Date();
      twoWeeksAgo.setDate(twoWeeksAgo.getDate() - 14);
      
      const oneWeekAgo = new Date();
      oneWeekAgo.setDate(oneWeekAgo.getDate() - 7);
      
      const resolvedDate = new Date(oneWeekAgo);
      resolvedDate.setHours(resolvedDate.getHours() + 2);

      await this.createIncident({
        serviceId: services[0].id,
        title: "Increased API Response Times",
        description: "Some users experienced slower than normal API response times due to database query optimization issues.",
        status: "resolved",
        severity: "minor",
        startedAt: twoWeeksAgo,
        resolvedAt: resolvedDate,
      });

      await this.createIncident({
        serviceId: services[1].id,
        title: "Storage Upload Failures",
        description: "A subset of object storage uploads failed due to temporary network issues in ru-central1-b region.",
        status: "resolved",
        severity: "major",
        startedAt: oneWeekAgo,
        resolvedAt: new Date(oneWeekAgo.getTime() + 3600000),
      });

      // GitLab incidents
      const gitlabService = services.find(s => s.name.includes("GitLab"));
      if (gitlabService) {
        const twoDaysAgo = new Date(Date.now() - 2 * 24 * 60 * 60 * 1000);
        const threeDaysAgo = new Date(Date.now() - 3 * 24 * 60 * 60 * 1000);
        const fiveDaysAgo = new Date(Date.now() - 5 * 24 * 60 * 60 * 1000);
        
        await this.createIncident({
          serviceId: gitlabService.id,
          title: "CI/CD Pipeline Timeout",
          description: "Build pipeline exceeded timeout limit",
          status: "resolved",
          severity: "minor",
          startedAt: twoDaysAgo,
          resolvedAt: new Date(twoDaysAgo.getTime() + 1800000),
        });

        await this.createIncident({
          serviceId: gitlabService.id,
          title: "Repository Sync Delay",
          description: "Git push operations experiencing delays",
          status: "resolved",
          severity: "major",
          startedAt: threeDaysAgo,
          resolvedAt: new Date(threeDaysAgo.getTime() + 2700000),
        });

        await this.createIncident({
          serviceId: gitlabService.id,
          title: "Merge Request Processing",
          description: "Automated merge checks running slower than expected",
          status: "resolved",
          severity: "minor",
          startedAt: fiveDaysAgo,
          resolvedAt: new Date(fiveDaysAgo.getTime() + 3600000),
        });

        await this.createIncident({
          serviceId: gitlabService.id,
          title: "Runner Connectivity",
          description: "GitLab runners intermittently disconnecting",
          status: "investigating",
          severity: "major",
          startedAt: new Date(Date.now() - 3 * 60 * 60 * 1000),
        });
      }

      // S3 Storage incidents
      const s3Service = services.find(s => s.name.includes("S3"));
      if (s3Service) {
        const oneDayAgo = new Date(Date.now() - 1 * 24 * 60 * 60 * 1000);
        const twoDaysAgo = new Date(Date.now() - 2 * 24 * 60 * 60 * 1000);
        const threeDaysAgo = new Date(Date.now() - 3 * 24 * 60 * 60 * 1000);
        const fourDaysAgo = new Date(Date.now() - 4 * 24 * 60 * 60 * 1000);
        const fiveDaysAgo = new Date(Date.now() - 5 * 24 * 60 * 60 * 1000);
        const sixDaysAgo = new Date(Date.now() - 6 * 24 * 60 * 60 * 1000);
        const sevenDaysAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);

        await this.createIncident({
          serviceId: s3Service.id,
          title: "High Latency on S3 Reads",
          description: "Object retrieval experiencing increased latency. Bucket size: 450GB",
          status: "resolved",
          severity: "major",
          startedAt: oneDayAgo,
          resolvedAt: new Date(oneDayAgo.getTime() + 4500000),
        });

        await this.createIncident({
          serviceId: s3Service.id,
          title: "Bucket Quota Warning",
          description: "Storage bucket approaching capacity limit. Size: 780GB/1TB",
          status: "monitoring",
          severity: "minor",
          startedAt: fourDaysAgo,
        });

        await this.createIncident({
          serviceId: s3Service.id,
          title: "Failed Upload Operations",
          description: "Multiple upload failures due to network congestion. Bucket: 320GB",
          status: "resolved",
          severity: "critical",
          startedAt: sixDaysAgo,
          resolvedAt: new Date(sixDaysAgo.getTime() + 5400000),
        });

        await this.createIncident({
          serviceId: s3Service.id,
          title: "S3 Replication Delay",
          description: "Cross-region replication experiencing delays. Bucket size: 125GB",
          status: "resolved",
          severity: "minor",
          startedAt: twoDaysAgo,
          resolvedAt: new Date(twoDaysAgo.getTime() + 2100000),
        });

        await this.createIncident({
          serviceId: s3Service.id,
          title: "Bucket Access Permissions Issue",
          description: "IAM policy conflicts blocking access. Affected bucket: 92GB",
          status: "resolved",
          severity: "major",
          startedAt: threeDaysAgo,
          resolvedAt: new Date(threeDaysAgo.getTime() + 3300000),
        });

        await this.createIncident({
          serviceId: s3Service.id,
          title: "Storage Class Migration",
          description: "Automated archival to Glacier delayed. Data volume: 1.2TB",
          status: "resolved",
          severity: "minor",
          startedAt: fiveDaysAgo,
          resolvedAt: new Date(fiveDaysAgo.getTime() + 7200000),
        });

        await this.createIncident({
          serviceId: s3Service.id,
          title: "Object Versioning Overflow",
          description: "Excessive versions consuming storage. Bucket size: 560GB",
          status: "resolved",
          severity: "major",
          startedAt: sevenDaysAgo,
          resolvedAt: new Date(sevenDaysAgo.getTime() + 4800000),
        });
      }
      
      // Create server metrics for the last 24 hours
      for (let i = 24; i >= 0; i--) {
        const timestamp = new Date();
        timestamp.setHours(timestamp.getHours() - i);
        
        for (const service of services) {
          const baseLoad = Math.random() * 40 + 30;
          
          // S3 services have higher disk usage
          const isS3 = service.name.toLowerCase().includes('s3') || service.name.toLowerCase().includes('storage');
          const diskUsage = isS3 ? 60 + Math.random() * 25 : 45 + Math.random() * 15;
          
          await this.createServerMetrics({
            serviceId: service.id,
            cpuUsage: baseLoad + Math.random() * 20,
            ramUsage: baseLoad + Math.random() * 25,
            diskUsage: diskUsage,
          });
        }
      }
    }
  }

  // Services
  async getServices(): Promise<Service[]> {
    return Array.from(this.services.values());
  }

  async getService(id: string): Promise<Service | undefined> {
    return this.services.get(id);
  }

  async createService(insertService: InsertService): Promise<Service> {
    const id = this.generateDeterministicId(insertService);
    const service: Service = {
      id,
      name: insertService.name,
      description: insertService.description ?? null,
      category: insertService.category,
      region: insertService.region,
      status: insertService.status || "operational",
      type: insertService.type ?? null,
      icon: insertService.icon ?? null,
      address: insertService.address ?? null,
      port: insertService.port ?? null,
      updatedAt: new Date(),
    };
    this.services.set(id, service);
    
    // Create initial status history entry
    await this.createStatusHistory({
      serviceId: id,
      status: insertService.status || "operational",
      timestamp: new Date(),
    });
    
    return service;
  }

  private generateDeterministicId(service: InsertService): string {
    const key = `${service.name}-${service.region}-${service.category}-${service.address || ''}-${service.port || ''}`;
    let hash = 0;
    for (let i = 0; i < key.length; i++) {
      const char = key.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    const hashStr = Math.abs(hash).toString(16).padStart(30, '0');
    return `${hashStr.slice(0,8)}-${hashStr.slice(8,12)}-4${hashStr.slice(12,15)}-a${hashStr.slice(15,18)}-${hashStr.slice(18,30)}`;
  }

  async updateServiceStatus(id: string, status: ServiceStatus): Promise<Service | undefined> {
    const service = this.services.get(id);
    if (!service) return undefined;

    const updatedService: Service = {
      ...service,
      status,
      updatedAt: new Date(),
    };
    this.services.set(id, updatedService);
    
    // Record status change in history
    await this.createStatusHistory({
      serviceId: id,
      status,
      timestamp: new Date(),
    });
    
    return updatedService;
  }

  // Incidents
  async getIncidents(): Promise<Incident[]> {
    return Array.from(this.incidents.values());
  }

  async getIncident(id: string): Promise<Incident | undefined> {
    return this.incidents.get(id);
  }

  async createIncident(insertIncident: InsertIncident): Promise<Incident> {
    const id = randomUUID();
    const incident: Incident = {
      ...insertIncident,
      id,
      startedAt: insertIncident.startedAt || new Date(),
      createdAt: new Date(),
    };
    this.incidents.set(id, incident);
    return incident;
  }

  // Status History
  async getStatusHistory(serviceId: string): Promise<StatusHistory[]> {
    return Array.from(this.statusHistory.values())
      .filter((history) => history.serviceId === serviceId)
      .sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
  }

  async createStatusHistory(insertHistory: InsertStatusHistory): Promise<StatusHistory> {
    const id = randomUUID();
    const history: StatusHistory = {
      ...insertHistory,
      id,
      timestamp: insertHistory.timestamp || new Date(),
    };
    this.statusHistory.set(id, history);
    return history;
  }

  // Server Metrics
  async getServerMetrics(serviceId?: string): Promise<ServerMetrics[]> {
    let metrics = Array.from(this.serverMetrics.values());
    if (serviceId) {
      metrics = metrics.filter((m) => m.serviceId === serviceId);
    }
    return metrics.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
  }

  async createServerMetrics(insertMetrics: InsertServerMetrics): Promise<ServerMetrics> {
    const id = randomUUID();
    const metrics: ServerMetrics = {
      ...insertMetrics,
      id,
      timestamp: new Date(),
    };
    this.serverMetrics.set(id, metrics);
    return metrics;
  }
}

export const storage = new MemStorage();

File: statuserver/server_py/storage.py
import uuid
import json
from datetime import datetime, timedelta
from typing import Dict, List, Optional
from pathlib import Path

from models import (
    Service, InsertService,
    Incident, InsertIncident,
    StatusHistory, InsertStatusHistory,
    ServerMetrics, InsertServerMetrics,
    ServiceStatus
)

class MemStorage:
    def __init__(self):
        self.services: Dict[str, Service] = {}
        self.incidents: Dict[str, Incident] = {}
        self.status_history: Dict[str, StatusHistory] = {}
        self.server_metrics: Dict[str, ServerMetrics] = {}
        
    async def seed_data(self):
        from import_data import import_services_from_data
        
        imported_file = Path(__file__).parent.parent / "server" / "imported_services.json"
        if imported_file.exists():
            with open(imported_file, 'r', encoding='utf-8') as f:
                imported_data = json.load(f)
            await import_services_from_data(self, imported_data)
    
    def _generate_deterministic_id(self, service: InsertService) -> str:
        key = f"{service.name}-{service.region}-{service.category}-{service.address or ''}-{service.port or ''}"
        hash_val = 0
        for char in key:
            hash_val = ((hash_val << 5) - hash_val) + ord(char)
            hash_val = hash_val & 0xFFFFFFFF
        
        hash_str = format(abs(hash_val), '030x')
        return f"{hash_str[0:8]}-{hash_str[8:12]}-4{hash_str[12:15]}-a{hash_str[15:18]}-{hash_str[18:30]}"
    
    async def get_services(self) -> List[Service]:
        return list(self.services.values())
    
    async def get_service(self, service_id: str) -> Optional[Service]:
        return self.services.get(service_id)
    
    async def create_service(self, insert_service: InsertService) -> Service:
        service_id = self._generate_deterministic_id(insert_service)
        service = Service(
            id=service_id,
            name=insert_service.name,
            description=insert_service.description,
            category=insert_service.category,
            region=insert_service.region,
            status=insert_service.status or "operational",
            type=insert_service.type,
            icon=insert_service.icon,
            address=insert_service.address,
            port=insert_service.port,
            updated_at=datetime.now()
        )
        self.services[service_id] = service
        
        await self.create_status_history(InsertStatusHistory(
            service_id=service_id,
            status=service.status,
            timestamp=datetime.now()
        ))
        
        return service
    
    async def update_service_status(self, service_id: str, status: ServiceStatus) -> Optional[Service]:
        service = self.services.get(service_id)
        if not service:
            return None
        
        updated_service = Service(
            **service.model_dump(),
            status=status,
            updated_at=datetime.now()
        )
        self.services[service_id] = updated_service
        
        await self.create_status_history(InsertStatusHistory(
            service_id=service_id,
            status=status,
            timestamp=datetime.now()
        ))
        
        return updated_service
    
    async def get_incidents(self) -> List[Incident]:
        return list(self.incidents.values())
    
    async def get_incident(self, incident_id: str) -> Optional[Incident]:
        return self.incidents.get(incident_id)
    
    async def create_incident(self, insert_incident: InsertIncident) -> Incident:
        incident_id = str(uuid.uuid4())
        incident = Incident(
            id=incident_id,
            service_id=insert_incident.service_id,
            title=insert_incident.title,
            description=insert_incident.description,
            status=insert_incident.status,
            severity=insert_incident.severity,
            started_at=insert_incident.started_at or datetime.now(),
            resolved_at=insert_incident.resolved_at,
            created_at=datetime.now()
        )
        self.incidents[incident_id] = incident
        return incident
    
    async def get_status_history(self, service_id: str) -> List[StatusHistory]:
        history = [h for h in self.status_history.values() if h.service_id == service_id]
        return sorted(history, key=lambda h: h.timestamp, reverse=True)
    
    async def create_status_history(self, insert_history: InsertStatusHistory) -> StatusHistory:
        history_id = str(uuid.uuid4())
        history = StatusHistory(
            id=history_id,
            service_id=insert_history.service_id,
            status=insert_history.status,
            timestamp=insert_history.timestamp or datetime.now()
        )
        self.status_history[history_id] = history
        return history
    
    async def get_server_metrics(self, service_id: Optional[str] = None) -> List[ServerMetrics]:
        metrics = list(self.server_metrics.values())
        if service_id:
            metrics = [m for m in metrics if m.service_id == service_id]
        return sorted(metrics, key=lambda m: m.timestamp, reverse=True)
    
    async def create_server_metrics(self, insert_metrics: InsertServerMetrics) -> ServerMetrics:
        metrics_id = str(uuid.uuid4())
        metrics = ServerMetrics(
            id=metrics_id,
            service_id=insert_metrics.service_id,
            cpu_usage=insert_metrics.cpu_usage,
            ram_usage=insert_metrics.ram_usage,
            disk_usage=insert_metrics.disk_usage,
            timestamp=datetime.now()
        )
        self.server_metrics[metrics_id] = metrics
        return metrics

storage = MemStorage()
